File: Assets/Framework/Entitas/Entitas/Systems/EventSystems.cs
///----------------------------------------------------------------
///<summary>
/// 版 本：v1.0.0
/// 创建人：Lonely
/// 日 期：2020/11/4 15:56:27
/// 描 述：
///</summary>
///----------------------------------------------------------------

using Entitas;
using System;
using System.Collections.Generic;
using UnityEngine;
using System.Text;


public class EventSystems : Feature
{
    /// <summary>
    /// Key:模块ID
    /// Value:
    /// {
    ///     Key:事件ID
    ///     Value:事件监听系统
    /// }
    /// </summary>
    protected readonly Dictionary<int, Dictionary<int, IEventSystem>> _event_systems;

    public EventSystems():base("Event Systems")
    {
        _event_systems = new Dictionary<int, Dictionary<int, IEventSystem>>();
    }

    protected void _registered(IEventSystem eventSystem)
    {
        int moduleID = eventSystem.GetModuleID();
        if (!_event_systems.ContainsKey(moduleID))
            _event_systems.Add(moduleID, new Dictionary<int, IEventSystem>());

        int[] listeners = eventSystem.GetListener();
        Dictionary<int, IEventSystem> datas = _event_systems[moduleID];

        for (int i = 0; i < listeners.Length; i++)
        {
            if (!datas.ContainsKey(listeners[i]))
                datas.Add(listeners[i], eventSystem);
        }
    }

    protected void _uninstall(IEventSystem eventSystem)
    {
        int moduleID = eventSystem.GetModuleID();
        if (!_event_systems.ContainsKey(moduleID))
            return;

        int[] listeners = eventSystem.GetListener();
        Dictionary<int, IEventSystem> datas = _event_systems[moduleID];
        for (int i = 0; i < listeners.Length; i++)
        {
            if (datas.ContainsKey(listeners[i]))
                datas.Remove(listeners[i]);
        }
    }

    protected void _notify(int moduleID, int op, System.Object param = null)
    {
        if (!_event_systems.ContainsKey(moduleID))
        {
            Debug.LogError(string.Format("模块=[{0}]没有注册事件", moduleID));
            return;
        }

        Dictionary<int, IEventSystem> events = _event_systems[moduleID];

        if (!events.ContainsKey(op))
        {
            Debug.LogError(string.Format("模块=[{0}]没有注册事件=[{1}]", moduleID, op));
            return;
        }

        events[op].Notify(op, param);
    }


}

File: Assets/L_Lua/Framework/entitas/EventSystems.lua
-- 事件系统
local M = class("EventSystems");

function  M:ctor()
	self._event_systems = {}
end

function M:Has( moduleID )
	 return self._event_systems[moduleID];
end

function M:_registered( eventSystem )
	local moduleID = eventSystem.GetModuleID();
	if not self:Has(moduleID) then
		self._event_systems[moduleID] = {}
	end

     local listeners = eventSystem.GetListener();

     for k,v in pairs(listeners) do
     	if v then
     		self._event_systems[moduleID][v] = eventSystem
     	end
     end

end

function M:_uninstall( eventSystem )
	local moduleID = eventSystem.GetModuleID();
	if not self:Has(moduleID) then
		return;
	end

	local listeners = eventSystem.GetListener();
    for k,v in pairs(listeners) do
     	if v then
     		self._event_systems[moduleID][v] = nil
     	end
     end

end

function M:_notify( moduleID,op,param)
	if not self:Has(moduleID) then
		error("该模块=["..moduleID.."]没有注册事件")
		return;
	end

	local tableList = self._event_systems[moduleID]

	if tableList[op] then
		tableList[op]:Notify(op,param)
	else
		error("模块=["..moduleID.."]不存在事件=["..op.."]")
	end
	
end


return M;